### PR TITLE
draft for creating a stripe ACH charge

### PR DESCRIPTION
**Pull request by jsun1**
[Do not merge yet]
@kiranmuddam This is draft code for how I believe the backend would initiate the ACH charge. (The code still needs some debugging before it can work). 

Requirements:
- `npm install stripe`
- Log into stripe.com and create a new API secret. Add your API secret to your .env file and call it `STRIPE_SECRET`

### CODE REVIEW

File: node/index.js [hunk 1]
```diff
@@ -222,6 +224,33 @@ app.post('/api/bank_account_token', function (request, response, next) {
...
 });
 
+app.post('/api/charge_ach', function (request, response, next) {
+  let bank_account_token = request.body.bank_account_token;
+  let amount_cents = request.body.amount; // Amount in cents
+  Promise.resolve()
+    .then(async function () {
+      const customer = await stripe(STRIPE_SECRET).customers.create({
+        source: bank_account_token,
+      });
```
Comment (jsun1) [inline node/index.js:234]: Currently this is returning `undefined`, so that would need to be fixed

**Review by kiranmuddam — APPROVED** [review]

**Comment by kiranmuddam** [comment]
@jsun1 I have fixed the issue and there is no error from the server side. But when I am testing in plaid sandbox env, using stripe testing SECRET_KEY. The response for the charge is Error: Token not found. I attached the error screenshot from stripe. After going through some research I am thinking the issue is with environment matching between plaid and stripe (or) problem with integration. Can you please confirm from the ios side if you are using any non-testing keys /  or you know specific thing that is causing this issue.
Screenshot 2022-08-25 at 4 50 19 PM

**Comment by jsun1** [comment]
Thanks Kiran, I'm also getting the same error as you now, and I'm using the production API keys for Plaid and Stripe:

```
Error: No such token: 'btok...'
    at Function.generate (/Users/jeffreysun/Documents/Candle/quickstart/node/node_modules/stripe/lib/Error.js:40:16)
    at res.toJSON.then.StripeAPIError.message (/Users/jeffreysun/Documents/Candle/quickstart/node/node_modules/stripe/lib/StripeResource.js:238:35)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

**Comment by jsun1** [comment]
cc @dragonmaster-alpha for context on the stripe integration for deposits

**Comment by jsun1** [comment]
Based on this stack overflow post, it seems perhaps we need to use the right combination of Plaid/Stripe keys: https://stackoverflow.com/questions/50286492/stripe-stripeinvalidrequesterror-error-no-such-token-btok-1cqutni6ucep4uarens

I'll try some of the suggestions from this post

**Comment by jsun1** [comment]
@samisbakedham Just to confirm, the Stripe account that's [connected to Plaid](https://dashboard.plaid.com/team/integrations) is the same one as the Stripe account that Kiran and I are added to?

Screen Shot 2022-08-25 at 7 24 48 PM

**Comment by samisbakedham** [comment]
> @samisbakedham Just to confirm, the Stripe account that's [connected to Plaid](https://dashboard.plaid.com/team/integrations) is the same one as the Stripe account that Kiran and I are added to?
> 
> 
> 
> Screen Shot 2022-08-25 at 7 24 48 PM
> 
> 

You know what, that's probably the issue

**Comment by jsun1** [comment]
Ah great thanks for adding the new Stripe account, Sam - I tested now and I do see the bank token is accepted and the server creates the customer object using it:

```
charge_ach customer: {
  id: 'cus_MJX...NyY',
  object: 'customer',
  address: null,
  balance: 0,
  created: 1661489362,
  currency: null,
  default_currency: null,
  default_source: 'ba_1Lau...nmFR',
  delinquent: false,
  description: 'test',
  discount: null,
  email: null,
  invoice_prefix: '991B0CEB',
  invoice_settings: {
    custom_fields: null,
    default_payment_method: null,
    footer: null,
    rendering_options: null
  },
  livemode: true,
  metadata: {},
  name: null,
  next_invoice_sequence: 1,
  phone: null,
  preferred_locales: [],
  shipping: null,
  tax_exempt: 'none',
  test_clock: null
}
```

**Comment by jsun1** [comment]
Cool, I just pushed a bug fix to master, and now passing the customer object to create an ACH charge appears to work. I transferred 50 cents twice from my bank account. This is the charge object returned:

```
charge_ach charge: {
  id: 'py_1Laup...AZL2',
  object: 'charge',
  amount: 50,
  amount_captured: 50,
  amount_refunded: 0,
  application: null,
  application_fee: null,
  application_fee_amount: null,
  balance_transaction: 'txn_1LaupC...TR888',
  billing_details: {
    address: {
      city: null,
      country: null,
      line1: null,
      line2: null,
      postal_code: null,
      state: null
    },
    email: null,
    name: null,
    phone: null
  },
  calculated_statement_descriptor: null,
  captured: true,
  created: 1661490682,
  currency: 'usd',
  customer: 'cus_MJXv...HJnz',
  description: null,
  destination: null,
  dispute: null,
  disputed: false,
  failure_balance_transaction: null,
  failure_code: null,
  failure_message: null,
  fraud_details: {},
  invoice: null,
  livemode: true,
  metadata: {},
  on_behalf_of: null,
  order: null,
  outcome: {
    network_status: 'approved_by_network',
    reason: null,
    risk_level: 'not_assessed',
    seller_message: 'Payment complete.',
    type: 'authorized'
  },
  paid: false,
  payment_intent: null,
  payment_method: null,
  payment_method_details: {
    ach_debit: {
      account_holder_type: null,
      bank_name: 'BANK OF AMERICA, N.A. (SFNB)',
      country: 'US',
      fingerprint: '5Gr...5U3',
      last4: '1407',
      routing_number: '12...4'
    },
    type: 'ach_debit'
  },
  receipt_email: null,
  receipt_number: null,
  receipt_url: null,
  refunded: false,
  refunds: {
    object: 'list',
    data: [],
    has_more: false,
    total_count: 0,
    url: '/v1/charges/py_1LaupCD...dOAZL2/refunds'
  },
  review: null,
  shipping: null,
  source: {
    id: 'ba_1Laup5D...W1p0r1a',
    object: 'bank_account',
    account_holder_name: null,
    account_holder_type: null,
    account_type: null,
    bank_name: 'BANK OF AMERICA, N.A. (SFNB)',
    country: 'US',
    currency: 'usd',
    customer: 'cus_MJXv...HJnz',
    fingerprint: '5Grw...5U3',
    last4: '1234',
    metadata: {},
    routing_number: '12...4',
    status: 'verified'
  },
  source_transfer: null,
  statement_descriptor: null,
  statement_descriptor_suffix: null,
  status: 'pending',
  transfer_data: null,
  transfer_group: null
}
```

**Comment by jsun1** [comment]
I've deployed the latest changes to api.maltafinance.org. Closing this PR now.

**Comment by dragonmaster-alpha** [comment]
Now can I use deployed api server rather than the local server?

**Comment by jsun1** [comment]
Yes, you can use the deployed api server now

**Comment by dragonmaster-alpha** [comment]
thanks for your confirmation

**Pull request closed without merging**